Add open API to allow rest framework to be documented.

rolfyone opened the pull request:
#1157 Add OpenApi to allow documentation of Rest Api

Add openAPI annotations to /node/genesisTime

Add /node/version endpoint with OpenApi annotations

rolfyone committed:
[WIP] add libraries required to get swagger displaying.

Restructured GenesisTimeHandler to allow custom Summary, Description, tags (grouping)

for some reason the licence string for MIT was not the same as other libraries, updated the licence check script.

Refactored BeaconRestApi class to be singleton, which allows us to follow the standard context pattern when we build the API endpoints.

rolfyone committed:
[WIP] put some basic structure around adding different handlers. also fix openapi description in genesis_time.

rolfyone committed:
update test cases, refactor a little after chat with Adrian.

cemozerr commented on data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApi.java:
```diff
@@ -75,6 +80,70 @@ public void start() {
+  }
+
+  private void addBeaconHandlers() {
+    // TODO: not in Minimal or optional specified set - some are similar to lighthouse
```
Very recently I've added beaconChainHead and modified beaconBlock to expose Alethio a similar API to Prysm's. FYI. It would be nice to connect you guys actually (with Alethio) since they are going to be using this API a lot.

rolfyone replied:
absolutely! Arash is coordinating as PO, we should get together and understand where things are at.

rolfyone committed:
add version information for openApi to use, implement VersionHandler to show the version string on request and associated test case.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into openApi

rolfyone committed:
added version prefix to genesis_time call, which is a breaking change and shown in acceptance test.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Add test to show that Javalin is created and that the GenesisTimeHandler and VersionHandler are added to that Javalin instance.

This is sufficient to show that endpoints are linked correctly to Javalin, and there are already at least some acceptance tests calling some endpoints, so the test gap was basically that we didn't show the endpoints get linked to Javalin on a specific route.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into openApi

rolfyone committed:
remove /v1 prefix, versioning would be via adding /v2 later for specific endpoints

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/GenesisTimeHandler.java:
```diff
@@ -13,24 +13,49 @@
   @Override
-  public Object handleRequest(RequestParams params) {
-    return client.getGenesisTime();
+  public void handle(@NotNull Context ctx) throws Exception {
```
nit: We don't typically use the `@NotNull` annotations (IntelliJ does sneak them in from time to time). They are unfortunately on the class path because of Kotlin dependencies.

ajsutton commented on data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApiTest.java:
```diff
@@ -0,0 +1,59 @@
+class BeaconRestApiTest {
+  private final ChainStorageClient storageClient =
+      ChainStorageClient.memoryOnlyClient(new EventBus());
+  private final JavalinServer mockServer = Mockito.mock(JavalinServer.class);
```
nit: Use a static import for `mock`.

ajsutton commented on data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApiTest.java:
```diff
@@ -0,0 +1,59 @@
+    when(mockApp.server()).thenReturn(mockServer);
+    new BeaconRestApi(storageClient, null, null, THE_PORT, mockApp);
+
+    Mockito.verify(mockApp).get(eq(VersionHandler.ROUTE), any(VersionHandler.class));
```
nit: static import for `verify`.

ajsutton commented on data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/GenesisTimeHandlerTest.java:
```diff
@@ -13,30 +13,37 @@
+    GenesisTimeHandler subject = new GenesisTimeHandler(null);
+    subject.handle(mockContext);
+
+    Mockito.verify(mockContext).status(SC_INTERNAL_SERVER_ERROR);
```
nit: static import :)

ajsutton commented on data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/GenesisTimeHandlerTest.java:
```diff
@@ -13,30 +13,37 @@
-  public void shouldReturnEmptyObjectWhenGenesisTimeIsNotSet() {
-    assertThat(handler.handleRequest(requestParams)).isEqualTo(null);
+  public void shouldRaiseInternalErrorWhenGenesisTimeIsNotSet() throws Exception {
+    GenesisTimeHandler subject = new GenesisTimeHandler(null);
```
nit: We rarely use names like `subject` for test subjects in the codebase - typically this would just be `handler`.

ajsutton commented on data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/GenesisTimeHandlerTest.java:
```diff
@@ -13,30 +13,37 @@
   @Test
-  public void shouldReturnEmptyObjectWhenGenesisTimeIsNotSet() {
-    assertThat(handler.handleRequest(requestParams)).isEqualTo(null);
+  public void shouldRaiseInternalErrorWhenGenesisTimeIsNotSet() throws Exception {
```
I'm not sure this should be an internal server error.  There is a, potentially long, period where a node is running prior to the genesis event occurring and we should have some meaningful response during that time (it may still be an error, but an expected one, not an internal server error).  404 Not found is probably the most accurate in REST terms but likely to be confusing for users who may think they got the URL wrong.

During that period we would have a `storageClient` but it's `store` and `genesisTime` fields would be `null`.

ajsutton reviewed (approved):
LGTM. Mostly just a bunch of little nits just around consistency with the rest of the codebase.

rolfyone committed:
changes per review comments.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
started a readme to try to capture any decisions we're making and store references to important resources.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge branch 'master' into openApi

macfarla commented on data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApi.java:
```diff
@@ -77,6 +94,69 @@ public void start() {
+    handlers.add(new BeaconChainHeadHandler(chainStorageClient));
+    handlers.add(new BeaconStateHandler(chainStorageClient));
+    handlers.add(new FinalizedCheckpointHandler(chainStorageClient));
+    handlers.add(new BeaconBlockHandler(chainStorageClient, historicalChainData));
```
nit: any reason for this ordering? why aren't all the beacon ones together

rolfyone replied:
these are all slated for deletion by the time we're done... I will re-order them if you prefer, but they're on borrowed time.

macfarla commented on data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApi.java:
```diff
@@ -77,6 +94,69 @@ public void start() {
+    // not in Minimal or optional specified set
+    handlers.add(new PeerIdHandler(p2pNetwork));
+    handlers.add(new PeersHandler(p2pNetwork));
+    handlers.add(new ENRHandler());
```
nit: EnrHandler ?

rolfyone replied:
leaving for this PR, per discussion.

macfarla commented on data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/GenesisTimeHandlerTest.java:
```diff
@@ -13,30 +13,38 @@
   @Test
-  public void shouldReturnEmptyObjectWhenGenesisTimeIsNotSet() {
-    assertThat(handler.handleRequest(requestParams)).isEqualTo(null);
+  public void shouldReturnNoContentGenesisTimeIsNotSet() throws Exception {
```
```suggestion
  public void shouldReturnNoContentWhenGenesisTimeIsNotSet() throws Exception {
```

macfarla reviewed (approved):
couple of nits

rolfyone committed:
Update data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/GenesisTimeHandlerTest.java

Co-Authored-By: Sally MacFarlane <sally.macfarlane@consensys.net>

rolfyone committed:
changes from review comments.

Signed-off-by: Paul Harris <paul.harris@consensys.net>
